### DISCUSSION TITLE
Go 语言网络轮询器的实现原理 | Go 语言设计与实现 · /golang-netpoller

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#591] [Announcements]
https://draveness.me/golang-netpoller

 Go 语言网络轮询器的实现原理

**Comment by skyitachi** [comment]
epoll和goroutine结合的太精妙了，感谢博主的详细讲解

**Comment by Subfire** [comment]
5种I/O模型中咋没有同步?
凡是让用户线程阻塞(不管长短)的I/O方案都是同步I/O. 也就是说, 阻塞、非阻塞、信号驱动式都是同步I/O

**Reply by Lee-Land** [reply]
同步是指内核通知用户的是就绪事件而不是完成事件

**Comment by draveness** [comment]
> 5种I/O模型中咋没有同步?
> 凡是让用户线程阻塞(不管长短)的I/O方案都是同步I/O. 也就是说, 阻塞、非阻塞、信号驱动式都是同步I/O

你认为的五种模型是哪五种？

**Comment by ChangwangChen** [comment]
想问下博主网络轮询器是全局的吗？

**Comment by ks864148379** [comment]
> @ChangwangChen
> 想问下博主网络轮询器是全局的吗？

看起来应该是全局的，runtime/netpoll_epoll.go里面有一个epfd，netpoll()方法里面就是用了这个fd。

这个最好让博主确认下是不是这样哈

**Comment by cooper-dc** [comment]
博主，golang的netpoll有专门的线程来等待类似epoll的返回吗？因为调用epoll就会阻塞线程，goroutine就没法调度了

**Comment by cooper-dc** [comment]
> @wtoeasy
> 博主，golang的netpoll有专门的线程来等待类似epoll的返回吗？因为调用epoll就会阻塞线程，goroutine就没法调度了

看后面一节确认是在系统监控线程中完成

**Comment by ningmonguo** [comment]
事件准备好后加入toRun里面是sysmon执行的，sysmon执行在m0的g0上，那是不是toRun里面的goroutine最后是被放到了m0的p的队列里了的？

**Comment by draveness** [comment]
> 事件准备好后加入toRun里面是sysmon执行的，sysmon执行在m0的g0上，那是不是toRun里面的goroutine最后是被放到了m0的p的队列里了的？

~~m0 上是没有 p 的，所以这些 Goroutine 会被直接放到全局的 runq 上~~

sysmon 直接运行在线程上，所以这些 Goroutine 会被直接放到全局的 runq 上

https://github.com/golang/go/blob/39a5ee52b9b41b1e4f4cf821c78ef5b7be68d181/src/runtime/proc.go#L2912-L2919

---

*2020-11-06 UPDATES: 已在章节中加入说明*

**Comment by ningmonguo** [comment]
> @draveness
> > 事件准备好后加入toRun里面是sysmon执行的，sysmon执行在m0的g0上，那是不是toRun里面的goroutine最后是被放到了m0的p的队列里了的？
> 
> m0 上是没有 pp 的，所以这些 Goroutine 会被直接放到全局的 runq 上
> 
> https://github.com/golang/go/blob/39a5ee52b9b41b1e4f4cf821c78ef5b7be68d181/src/runtime/proc.go#L2912-L2919


我看着这里在启动的时候初始p数组的时候像是m0和p[0] 绑定关系的过程，不知道理解得对不对
https://github.com/golang/go/blob/39a5ee52b9b41b1e4f4cf821c78ef5b7be68d181/src/runtime/proc.go#L4751-L4755

**Comment by draveness** [comment]
> > @draveness
> > > 事件准备好后加入toRun里面是sysmon执行的，sysmon执行在m0的g0上，那是不是toRun里面的goroutine最后是被放到了m0的p的队列里了的？
> > 
> > 
> > m0 上是没有 pp 的，所以这些 Goroutine 会被直接放到全局的 runq 上
> > https://github.com/golang/go/blob/39a5ee52b9b41b1e4f4cf821c78ef5b7be68d181/src/runtime/proc.go#L2912-L2919
> 
> 我看着这里在启动的时候初始p数组的时候像是m0和p[0] 绑定关系的过程，不知道理解得对不对
> https://github.com/golang/go/blob/39a5ee52b9b41b1e4f4cf821c78ef5b7be68d181/src/runtime/proc.go#L4751-L4755

额，是 sysmon 不持有 P，之前的回复可能有歧义，看下 [6.7.2](https://draveness.me/golang/docs/part3-runtime/ch06-concurrency/golang-sysmon/#672-%E7%9B%91%E6%8E%A7%E5%BE%AA%E7%8E%AF)

> 系统监控的 Goroutine 会直接在创建的线程上运行

**Comment by ningmonguo** [comment]
> @draveness
> > > @draveness
> > > > 事件准备好后加入toRun里面是sysmon执行的，sysmon执行在m0的g0上，那是不是toRun里面的goroutine最后是被放到了m0的p的队列里了的？
> > > 
> > > 
> > > m0 上是没有 pp 的，所以这些 Goroutine 会被直接放到全局的 runq 上
> > > https://github.com/golang/go/blob/39a5ee52b9b41b1e4f4cf821c78ef5b7be68d181/src/runtime/proc.go#L2912-L2919
> > 
> > 我看着这里在启动的时候初始p数组的时候像是m0和p[0] 绑定关系的过程，不知道理解得对不对
> > https://github.com/golang/go/blob/39a5ee52b9b41b1e4f4cf821c78ef5b7be68d181/src/runtime/proc.go#L4751-L4755
> 
> 额，是 sysmon 不持有 P，之前的回复可能有歧义，看下 [6.7.2](https://draveness.me/golang/docs/part3-runtime/ch06-concurrency/golang-sysmon/#672-%E7%9B%91%E6%8E%A7%E5%BE%AA%E7%8E%AF)
> 
> > 系统监控的 Goroutine 会直接在创建的线程上运行

嗯嗯嗯，我的问题，我理解错了，sysmon不在m0

**Comment by Explorer0** [comment]
大佬，请问下netpoll到底如何监测到【定时器】的超时啊？
我看了【定时器】、【网络轮训器】2节，发现addtimer()时，写入netpollBreakWr，netpoll()读取netpollBreakRd并丢掉数据，啥也没做，中间没有定时器沉睡协程的出现

**Comment by draveness** [comment]
> 大佬，请问下netpoll到底如何监测到【定时器】的超时啊？
> 我看了【定时器】、【网络轮训器】2节，发现addtimer()时，写入netpollBreakWr，netpoll()读取netpollBreakRd并丢掉数据，啥也没做，中间没有定时器沉睡协程的出现

`netpoll` 读取 `netpollBreakRd` 的时候不就被唤醒了么

> 初始化的管道为我们提供了中断多路复用等待文件描述符中事件的方法，runtime.netpollBreak 函数会向管道中写入数据唤醒 epoll：

**Comment by luozhiyun993** [comment]
请问一下，是不是runtime.netpoll 主要有三处调用 1. findrunnable中 netpoll(0) 是非阻塞；2. findrunnable中 netpoll(delta) 阻塞调用 3. sysmon系统调用 netpoll(0) 是非阻塞；
所以 netpollBreak 主要唤醒的是上面第二处 findrunnable中 netpoll(delta) 阻塞调用 ？

**Comment by dchaofei** [comment]
```go
	var events [128]epollevent
retry:
	n := epollwait(epfd, &events[0], int32(len(events)), waitms)
```

这里最大扫描前128个fd,如果waitms等于-1,并且这128个fd一直没有就绪是不是程序就一直阻塞这里了？如果第129个fd可以用了，但是只扫描了128个，导致永远阻塞是不是有问题的？

博主能解答下疑惑吗，谢谢

**Comment by draveness** [comment]
> @dchaofei ```go
> 	var events [128]epollevent
> retry:
> 	n := epollwait(epfd, &events[0], int32(len(events)), waitms)
> ```
> 
> 这里最大扫描前128个fd,如果waitms等于-1,并且这128个fd一直没有就绪是不是程序就一直阻塞这里了？如果第129个fd可以用了，但是只扫描了128个，导致永远阻塞是不是有问题的？
> 
> 博主能解答下疑惑吗，谢谢

这里一次只支持获取 128 个事件，不是文件描述符个数

**Comment by dchaofei** [comment]
> @draveness
> > @dchaofei ```go
> > 	var events [128]epollevent
> > retry:
> > 	n := epollwait(epfd, &events[0], int32(len(events)), waitms)
> > ```
> > 
> > 这里最大扫描前128个fd,如果waitms等于-1,并且这128个fd一直没有就绪是不是程序就一直阻塞这里了？如果第129个fd可以用了，但是只扫描了128个，导致永远阻塞是不是有问题的？
> > 
> > 博主能解答下疑惑吗，谢谢
> 
> 这里一次只支持获取 128 个事件，不是文件描述符个数

感谢，是我对epollwait理解有误了
```
epoll_wait(int epfd, struct epoll_event *events, int maxevents, int timeout);
```
maxevents 是最大同时可处理的事件，如果epollwait里有200个事件可处理，只会返回128个，另外的会在下次调用 epoll_wait 时返回

**Comment by dchaofei** [comment]
> @Explorer0
> 大佬，请问下netpoll到底如何监测到【定时器】的超时啊？
> 我看了【定时器】、【网络轮训器】2节，发现addtimer()时，写入netpollBreakWr，netpoll()读取netpollBreakRd并丢掉数据，啥也没做，中间没有定时器沉睡协程的出现

这里列出 findrunnable() 部分代码
```go
top:
    ......
   now, pollUntil, _ := checkTimers(_p_, 0)  
   ......
   list := netpoll(delta)
		if faketime != 0 && list.empty() {
			// Using fake time and nothing is ready; stop M.
			// When all M's stop, checkdead will call timejump.
			stopm()
			goto top
		}
```

netpoll()读取netpollBreakRd并丢掉数据时说明netpoll()已经被中断了，程序会返回 findrunnable()函数里，如果netpoll返回的是空的g列表，那就会 goto top, 然后执行 checkTimer(), 这个函数里会唤醒到期的定时器。

**Comment by MrGuin** [comment]
只有平台支持的情况下，磁盘文件IO才能被注册到netpoll里的，至少Linux下是不行的。  
毕竟epoll只支持网络IO，不像kqueue可以支持文件IO。newFile初始化文件时，163行也会判断文件是否pollable，如果不是的话是无法注册到netpoller中的，注释里也说明了：
>// An error here indicates a failure to register  
// with the netpoll system. That can happen for  
// a file descriptor that is not supported by  
// epoll/kqueue; for example, disk files on  
// GNU/Linux systems.
  
大多数博客里直接把文件IO跟网络IO分开了，文件IO会阻塞线程，网络IO由epoll处理不会阻塞线程。  
我建议补充区分一下这一点，我就是看了很多博客，开始混乱到底文件IO会不会直接阻塞线程。

**Comment by Yarollingstone** [comment]
> @MrGuin
> 只有平台支持的情况下，磁盘文件IO才能被注册到netpoll里的，至少Linux下是不行的。  
> 毕竟epoll只支持网络IO，不像kqueue可以支持文件IO。newFile初始化文件时，163行也会判断文件是否pollable，如果不是的话是无法注册到netpoller中的，注释里也说明了：
> >// An error here indicates a failure to register  
> // with the netpoll system. That can happen for  
> // a file descriptor that is not supported by  
> // epoll/kqueue; for example, disk files on  
> // GNU/Linux systems.
>   
> 大多数博客里直接把文件IO跟网络IO分开了，文件IO会阻塞线程，网络IO由epoll处理不会阻塞线程。  
> 我建议补充区分一下这一点，我就是看了很多博客，开始混乱到底文件IO会不会直接阻塞线程。

感觉结合 epoll 的实现讲下调理会更清晰，epoll 是个池子，linux 底层用红黑树实现，epoll 池子负责 fd 的添加、查找、删除操作。epoll_wait 负责 fd 事件的阻塞式获取，只有实现了 file_operations 的 poll 接口的文件系统才能使用 epoll。

**Comment by Yarollingstone** [comment]
@draveness
博主，我看了下 函数 `poll_runtime_pollSetDeadline` 的源码，发现它里面对 pd 的操作都加了锁，既然都加了锁。这里 rd 和 wd 是什么时候设置的小于 0 的呢？是一开始传进来的 d 就是负数吗？
```go
var rg, wg *g
	if pd.rd < 0 || pd.wd < 0 {
		atomic.StorepNoWB(noescape(unsafe.Pointer(&wg)), nil) // full memory barrier between stores to rd/wd and load of rg/wg in netpollunblock
		if pd.rd < 0 {
			rg = netpollunblock(pd, 'r', false)
		}
		if pd.wd < 0 {
			wg = netpollunblock(pd, 'w', false)
		}
	}
	unlock(&pd.lock)
	if rg != nil {
		netpollgoready(rg, 3)
	}

**Comment by jwrookie** [comment]
感谢博主的详细讲解
```
func netpollunblock(pd *pollDesc, mode int32, ioready bool) *g {
	gpp := &pd.rg
	if mode == 'w' {
		gpp = &pd.wg
	}

	for {
		old := *gpp
		if old == pdReady {
			return nil
		}
		if old == 0 && !ioready {
			// Only set pdReady for ioready. runtime_pollWait
			// will check for timeout/cancel before waiting.
			return nil
		}
		var new uintptr
		if ioready {
			new = pdReady
		}
		if atomic.Casuintptr(gpp, old, new) {
			if old == pdWait {
				old = 0
			}
			return (*g)(unsafe.Pointer(old))
		}
	}
}
```
在获取被阻塞的g时，不太懂这里为什么要用for循环？

**Comment by tutengdihuang** [comment]
```
	pd := c.first
	c.first = pd.link
	unlock(&c.lock)
	return pd
```
- 这里少一句 初始化锁
```
	pd := c.first
	c.first = pd.link
	lockInit(&pd.lock, lockRankPollDesc)
	unlock(&c.lock)
	return pd
```

**Reply by xuxiao415** [reply]
请教一下，调用ioutil.readall读取http body体时rt_99很高会是什么原因呢？不知道和网络轮训器是否有关，我的golang 版本1.15.5

**Comment by overstep123** [comment]
网络轮询器不支持磁盘文件io的吧，epoll本身也不支持磁盘文件io，大佬你写的有验证过吗

**Reply by SuanCaiYv** [reply]
epoll确实不支持文件IO，即使是在io_uring依旧是阻塞式读写的，整个Linux都不支持文件IO的nonblocking。

**Comment by IguoChan** [comment]
网络轮询器不仅用于监控网络 I/O，还能用于监控文件的 I/O，它利用了操作系统提供的 I/O 多路复用模型来提升 I/O 设备的利用率以及程序的性能。
——以上这句的结论对么？貌似Linux下使用的epoll并不支持磁盘io的。

**Reply by SuanCaiYv** [reply]
监控文件IO指的是kqueue之类的，原文应该要做一下区分。

**Reply by IguoChan** [reply]
kqueue应该是MacOS中使用的吧，Linux是整体不支持的

**Comment by 454270186** [comment]
“runtime.netpoll 返回的 Goroutine 列表都会被 runtime.injectglist 注入到处理器或者全局的运行队列上。因为系统监控 Goroutine 直接运行在线程上，所以它获取的 Goroutine 列表会直接加入全局的运行队列，其他 Goroutine 获取的列表都会加入 Goroutine 所在处理器的运行队列上。“

在go1.23源码中，其他Goroutine获取到的列表不一定会加入其所在处理器的runq上，而是先判断全局有没有idle的p，如果有空闲的p，这个列表还是会被加入全局g队列。
https://github.com/golang/go/blob/master/src/runtime/proc.go#L4026
